### PR TITLE
Testing with pytest

### PR DESCRIPTION
**Pull request by B1T0**
I have created a new way of testing with pytest. 

A massively overengineered test class, `TestFunctional.py`, uses all options but is super inefficient due to combinatorial explosion.

A more simplistic script, `test_installation.py`, can be called from the base directory of Wochenende by activating the (modified) environment and executing `pytest`. It will test for all paths and report whether the pipeline can be executed correctly.

I tried to document functionality as good as possible, but it might still not be sufficient. Please ask questions to make it better.

### CODE REVIEW

File: env.wochenende.minimal.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: wochenende
+name: miniWE
```
Comment (colindaven) [inline env.wochenende.minimal.yml:1]: It is confusing to change the wochenende env name. We would need to change many scripts and docu to make this clear.

File: test/TestFunctional.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    # TODO: Sum up test run information
+
+    _p_fixed_arguments = [
+        '--metagenome', 'test',
+        '--debug'
+    ]
+    _p_options = {
+        '--aligner': ['bwamem', 'minimap2', 'ngmlr'],
+        '--readType': ['SE', 'PE'],
+        '--threads': ['8', '4', '8', '16', '32', '56', '80'],
```
Comment (colindaven) [inline test/TestFunctional.py:42]: There is repetition here. 8, 4, 8

File: test/TestFunctional.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    ]
+
+    def _gen_fixed_arguments(self):
+        """Returns the fixed arguments. Mainly for consistency."""
+        return self._p_fixed_arguments
+
+    def _gen_options(self):
+        """Generates all combinations of p_options
+
+        Please don't ask how. It works. Trust me. Sorry.
```
Comment (colindaven) [inline test/TestFunctional.py:65]: lol

File: test/TestFunctional.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+        print("# Options to test")
+        options = list(self._gen_options())
+        print(options)
+
+        print("# Flags to test")
+        flags = list(self._gen_flags())
+        # print(flags)
+
+        return [fixed + opts + flgs for opts, flgs in product(options, flags)]
```
Comment (colindaven) [inline test/TestFunctional.py:98]: is flgs a typo ?

File: test/TestFunctional.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        Returns
+        -------
+        refs : list of os.path
+            absolute paths to reference file(s)
+        reads : list of os.path
+            absolute paths to read file(s)
+
+        Note
+        ----
+        For testing, we provide explicit reference files. No reference from one of the
```
Comment (colindaven) [inline test/TestFunctional.py:112]: ok, but why ? Because user systems will have different paths  ?

File: test/TestFunctional.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        datadir = os.path.join(cwd, 'data')
+        datalist = os.listdir(datadir)
+
+        refs = [os.path.join(datadir, f) for f in datalist if f.endswith('.fa')]
+        reads = [os.path.join(datadir, f) for f in datalist if f.endswith('_R1.fastq')]
+
+        return refs, reads
+
+    def _get_fixed_files_paths(self):
+        """Returns singleton list of list of absolute paths of necessary files."""
```
Comment (colindaven) [inline test/TestFunctional.py:130]: example needed ? What is necessary ? For testing ? Wochenende ? Dependencies ?

File: test/test_Wochenende_SLURM.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+# set name of job
+#SBATCH --job-name=Wochenende_test
+
+# Add miniconda3 to PATH
+. /mnt/ngsnfs/tools/miniconda3/etc/profile.d/conda.sh
+
+# Activate env on cluster node
+conda activate wochenende >> /dev/null
+
+pytest --basetemp=/ngsssd1/rcug/tmp/ -n 4
```
Comment (colindaven) [inline test/test_Wochenende_SLURM.sh:27]: We probably should add a `srun -c $cpus pytest `...  and define `cpus=4` in the line above to the use the parameter in SLURM. 
`#SBATCH --cpus-per-task 4` is not sufficient weirdly.

**Review by colindaven — APPROVED** [review]
Thanks Tobias, this looks very nice. I think we should merge it in and test widely in dev before merging into master.

File: test/TestFunctional.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    # TODO: Test partially successful runs, continuation and --force-restart
+    # TODO: Collect test run information
+    # TODO: Sum up test run information
+
+    _p_fixed_arguments = [
+        '--metagenome', 'test',
+        '--debug'
+    ]
+    _p_options = {
+        '--aligner': ['bwamem', 'minimap2', 'ngmlr'],
```
Comment (colindaven) [inline test/TestFunctional.py:40]: there are now minimap2short and minimap2long as aligner. You'll need to add both here.

**Comment by poer-sophia** [comment]
I tried to run the scripts but when I run the test_insatllation.py I get this error.

File "test_installation.py", line 2, in <module>
    import run_Wochenende as we
ModuleNotFoundError: No module named 'run_Wochenende'

And when I run the test_Wochenende_SLURM skript I get this Error:
test_Wochenende_SLURM.sh: line 27: pytest: command not found

I ran both scripts with the modified wochenende enviroment activated. Did I oversee or forgot something? Do I need to install something else?

**Comment by colindaven** [comment]
@poer-sophia Thanks for testing. I'm finding it difficult to pull the PR and test it properly as well. But .. seeing as it is for devs, not users, do you think we can merge it in and continue this discussion with Tobias later?  As long as the basic functionality of Woch is untouched, it should be ok right?

Got another meeting now ..

**Comment by poer-sophia** [comment]
Yes thats right, it should be ok

**Pull request merged**